### PR TITLE
Unit test for pallet-staking unsorted bags feature (targets #9081)

### PR DESCRIPTION
**Pull request by emostov**
target branch: prgn-nominator-unsorted-bags
target pr:  #9081

cc @kianenigma

### CODE REVIEW

File: frame/staking/src/voter_bags.rs [hunk 1]
```diff
@@ -1024,23 +1037,132 @@ mod voter_list {
...
+
+			// then
+			assert_eq!(
+				VoterList::<Test>::iter().map(|n| n.voter().id).collect::<Vec<_>>(),
+				vec![21, 101, 31]
+			);
+			assert_eq!(get_bags(), vec![(10, vec![31]), (1_000, vec![21, 101])]);
+
+
+			// Remove from a bag with only one node:
```
Comment (emostov) [inline frame/staking/src/voter_bags.rs:1156]: is testing this scenario overkill?

File: frame/staking/src/voter_bags.rs [hunk 2]
```diff
@@ -960,7 +960,20 @@ mod voter_list {
 
 	#[test]
 	fn notional_bag_for_works() {
-		todo!();
+		// under a threshold gives the next threshold.
```
Comment (kianenigma) [inline frame/staking/src/voter_bags.rs:963]: nice 👍

File: frame/staking/src/mock.rs [hunk 3]
```diff
@@ -503,13 +503,21 @@ fn post_conditions() {
...
-	let nominator_count = Nominators::<Test>::iter().count() as u32;
-	let validator_count = Validators::<Test>::iter().count() as u32;
-	assert_eq!(nominator_count, CounterForNominators::<Test>::get());
-	assert_eq!(validator_count, CounterForValidators::<Test>::get());
+	// @kianenigma
+	// TODO: checking this is good for tests of top level staking api, but when
+	// unit testing parts of voter_bags we can't expect this to update properly
+	// (unless we manually do it in the test, which might be ok?).
+	// Instead I think the debug_asserts should be enough?
+	// Otherwise maybe we can have a `build_and_execute_without_check_count`, but that
```
Comment (kianenigma) [inline frame/staking/src/mock.rs:511]: `build_and_execute_without_check_count` sounds good, but use it only with extra attention, and it should ONLY be used in the `VoterList` api later or lower, not the top level tests.

File: frame/staking/src/voter_bags.rs [hunk 4]
```diff
@@ -1021,23 +1034,131 @@ mod voter_list {
...
-		todo!()
+		ExtBuilder::default().build_and_execute(|| {
+			// given
+			assert_eq!(
+				VoterList::<Test>::iter().map(|n| n.voter().id).collect::<Vec<_>>(),
+				vec![11, 21, 101, 31]
+			);
+			// TODO maybe checking the actual bags here is overkill since this is aimed at VoterList
+			// api and not bags? (same Q for other tests here)
+			assert_eq!(get_bags(), vec![(10, vec![31]), (1_000, vec![11, 21, 101])]);
```
Comment (kianenigma) [inline frame/staking/src/voter_bags.rs:1051]: I'm fine with testing as many stuff as possible. 

To help make it more tidy, maybe make a `fn iter_voter_list() -> Vec<AccountId>`.

Comment (kianenigma) [reply frame/staking/src/voter_bags.rs:1051]: Also as per the intention of this test module: 

```
// This is the highest level of abstraction provided by this module. More generic tests are here,
// among those related to `VoterList` struct.
#[cfg(test)]
mod voter_list { ... }
```

I think we should manually check all storage items. This is the same API that the top level pallet will use (e.g. `VoterList::insert`), it is important to see exactly what is in storage before and after an insert.

File: frame/staking/src/tests.rs [hunk 5]
```diff
@@ -3827,88 +3827,103 @@ fn on_finalize_weight_is_nonzero() {
...
+			// does not change the voters position
+			assert_eq!(
+				get_bags(),
+				vec![(10, vec![31]), (1000, vec![11, 21, 101]), (2000, vec![42])]
+			);
+
+			// `chill`
+			assert_ok!(Staking::chill(Origin::signed(43)));
+			// removes the voter
+			assert_eq!(get_bags(), vec![(10, vec![31]), (1000, vec![11, 21, 101]), (2000, vec![])]);
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:3865]: seems like we have a stale storage entry here, the 2000 bag which is empty. not a huge deal, but ideally once we remove the last item from the bag, it should be totally cleared.

File: frame/staking/src/tests.rs [hunk 6]
```diff
@@ -3827,88 +3827,103 @@ fn on_finalize_weight_is_nonzero() {
...
+			// increase stake and implicitly rebag with `bond_extra` to the level of non-existent bag
+			assert_ok!(Staking::bond_extra(Origin::signed(42), 1_980)); // 20 + 1_980 = 2_000
+			assert_eq!(
+				get_bags(),
+				vec![(10, vec![31]), (20, vec![]), (1000, vec![11, 21, 101]), (2000, vec![42])]
+			);
+
+			// decrease stake within the range of the current bag
+			assert_ok!(Staking::unbond(Origin::signed(43), 999)); // 2000 - 999 = 1001
+			assert_ok!(Staking::rebag(Origin::signed(43), 42));
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:3888]:
```suggestion
```

Comment (kianenigma) [reply frame/staking/src/tests.rs:3888]: unbond should automatically rebag itself afaik.

File: frame/staking/src/tests.rs [hunk 7]
```diff
@@ -3827,88 +3827,103 @@ fn on_finalize_weight_is_nonzero() {
...
+					(20, vec![]),
+					(30, vec![42]),
+					(1000, vec![11, 21, 101]),
+					(2000, vec![])
+				]
+			);
+
+			// increase stake by `rebond`-ing to the level of a pre-existing bag
+			assert_ok!(Staking::rebond(Origin::signed(43), 31)); // 30 + 41 = 61
+			assert_ok!(Staking::rebag(Origin::signed(43), 42));
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:3912]: again, should these not call rebag automatically?

Comment (kianenigma) [reply frame/staking/src/tests.rs:3912]: fixed it in [`e7fdad2` (#9422)](https://github.com/paritytech/substrate/pull/9422/commits/e7fdad2aa55c5e71ee209a11baf8b3014248f260)

File: frame/staking/src/tests.rs (hunk 5 above)
Comment (kianenigma) [reply frame/staking/src/tests.rs:3865]: fixed it in [`e7fdad2` (#9422)](https://github.com/paritytech/substrate/pull/9422/commits/e7fdad2aa55c5e71ee209a11baf8b3014248f260)

File: frame/staking/src/tests.rs (hunk 7 above)
Comment (emostov) [reply frame/staking/src/tests.rs]: 887035169fabb7778e806c8a6f36fa6c092fe5e2

File: frame/staking/src/tests.rs (hunk 6 above)
Comment (emostov) [reply frame/staking/src/tests.rs]: 887035169fabb7778e806c8a6f36fa6c092fe5e2

File: frame/staking/src/tests.rs (hunk 5 above)
Comment (emostov) [reply frame/staking/src/tests.rs]: remaining tests fixed in 724c17ef1bf4005a033c04ee1e5d8e4d79240714

**Comment by emostov** [comment]
@kianenigma   I updated `rebond` and `ubond` to `do_rebag` (which addreses your comments on this pr and [my comment on the feature pr](https://github.com/paritytech/substrate/pull/9081#issuecomment-887145748)) in this commit 887035169fabb7778e806c8a6f36fa6c092fe5e2

However `rebond` weights looks strange to me:

```
			Ok(Some(
				35 * WEIGHT_PER_MICROS +
					50 * WEIGHT_PER_NANOS * (ledger.unlocking.len() as Weight) +
					T::DbWeight::get().reads_writes(3, 2),
			)
```
and I am not sure how we should update it to account for rebag-ing? Should we update the benchmarks to force a rebag and then just use that weight? (I imagine we should update benchmarks for any extrinsic that does a rebag, or have rebag return a weight and use that)

**Comment by emostov** [comment]
@kianenigma https://github.com/paritytech/substrate/pull/9422/commits/ec7b7cc223ffeb634ae7cd357a1500fdcfd4a1e7 adds in a check to prevent the a scenario we know that could lead to infinite loop:

```rust
fn insert_node(&mut self, mut node: Node<T>) {
		if let Some(tail) = &self.tail {
			if *tail == node.voter.id {
				// this should never happen, but this check prevents a worst case infinite loop
				debug_assert!(false, "system logic error: inserting a node who has the id of tail");
				crate::log!(warn, "system logic error: inserting a node who has the id of tail");
				return
			};
		}
...
}
```

I think the argument against this would be that this should never happen. But considering that this check is 1) easy & cheap and 2) it would prevent the major issue of the runtime getting stuck on a loop it seems worth it from my PoV

**Comment by kianenigma** [comment]
> and I am not sure how we should update it to account for rebag-ing? Should we update the benchmarks to force a rebag and then just use that weight? 

Yes the benchmark need to be update to the worse case, which means doing a rebag as well. 

Once this is fixed, then we can merge this into the base branch

**Pull request merged**